### PR TITLE
CSE-921 export props and images

### PR DESCRIPTION
**Pull request by voodoods**
**Jira Issue**

https://8select.atlassian.net/browse/CSE-921

**Summary**

- replaced QuickUpdate Export with PropertyExport
- refactored Export logic so that shared code between `ArticleExport.php` and `PropertyExport.php` is bundled in `Export.php` parent class
- make the changes cronjob pickup changes to the configurator- and feature-option-tables
- make `CRON_NAME` in `ArticleExport.php` match the `name` of the cronjob that gets added to `s_crontab` DB table
- update `README.md` with new infos on usage with `copy2dev.sh`

**Checklist**

- [x] PR title is prefixed with Jira Issue Id - e.g. CSE-42
- [x] Add feature / bug label
- [x] Unit tests for new / changed logic exist OR no logic changes are passing

**Example export file:**
http://productfeed-prod.staging.8select.io.s3.amazonaws.com/150f0d73-ec87-4d25-8589-a2ac5a79be37/property_feed/2018/06/25/150f0d73-ec87-4d25-8589-a2ac5a79be37_property_feed_1529919053587.csv

### CODE REVIEW

**Comment by voodoods** [comment]
**Notes:**
* not pretty (as for now) but working
* the property export will pick up article-based deltas (will only include articles that have changed) but not value-based deltas (will include all available property values for the changed articles, regardless of which have changed and which not)
* code refactoring and value delta export will be handled in JIRA ticket

**Consideration:**
To make this feature more useful we should release it in combination with the multi-field-mapping in PR https://github.com/8select/shopware-plugin-sob/pull/17 
Currently we can only have a one-on-one mapping relation, so the property export can only pick up changes to this one specific field. 

**Example:** 
If you map the configurator option "Farbe / Material" to the field "FARBE" (color / prop_color):

bildschirmfoto 2018-06-25 um 11 42 07

only changes to the configurator options will be picked up by the plugin:

bildschirmfoto 2018-06-25 um 11 43 48

However if the user applies changes to the feature option "Farbe / Material" 

bildschirmfoto 2018-06-25 um 11 44 08

then those will not be picked up by the plugin, unless this field is also mapped to FARBE!

File: CseEightselectBasic/Components/PropertyExport.php [hunk 1]
```diff
@@ -96,16 +105,73 @@ public function doCron()
...
             RunCronOnce::finishCron(self::CRON_NAME);
             throw $exception;
         }
     }
 
+    public static function getOriginalFieldNames($fields) {
+
+        $origFields = array();
+
+        foreach ($fields as $field) {
```
Comment (rschweizer) [inline CseEightselectBasic/Components/PropertyExport.php]: I think usage of `array_map` would be clearer, e.g.:

```php
$origFields = array_map(
  function($field) {
    switch($field) {
      case 'prop_sku':
        return 'sku'
      case ...
         ...
    }
  }, 
  $fields
)
```

This way you could even simply `return array_map(..., $fields)`.

File: CseEightselectBasic/Components/PropertyExport.php [hunk 2]
```diff
@@ -154,6 +228,12 @@ protected function getArticles($from, $number)
                 INNER JOIN 8s_articles_details_change_queue ON 8s_articles_details_change_queue.s_articles_details_id = s_articles_details.id
                 INNER JOIN s_core_tax ON s_core_tax.id = s_articles.taxID
                 LIMIT ' . $number . ' OFFSET ' . $from;
+
+        if (getenv('ES_DEBUG')) {
```
Comment (rschweizer) [inline CseEightselectBasic/Components/PropertyExport.php]: What does `ES` stand for? Since you are only echoing SQL, should it not be `SQL_DEBUG`?

File: CseEightselectBasic/CseEightselectBasic.php [hunk 3]
```diff
@@ -822,11 +821,19 @@ private function initChangesQueueTable()
...
+                    IF (NEW.supplierID != OLD.supplierID OR NEW.configurator_set_id != OLD.configurator_set_id) THEN
+                      INSERT INTO 8s_articles_details_change_queue (s_articles_details_id, updated_at) VALUES (NEW.id, NOW());
+                    END IF;
+                  END',
             'DROP TRIGGER IF EXISTS `8s_articles_details_change_queue_writer`',
             'CREATE TRIGGER `8s_articles_details_change_queue_writer` AFTER UPDATE on `s_articles_details`
                   FOR EACH ROW
                   BEGIN
-                    IF (NEW.instock != OLD.instock OR NEW.active != OLD.active) THEN
+                    IF (NEW.ordernumber != OLD.ordernumber OR NEW.instock != OLD.instock OR NEW.active != OLD.active OR NEW.ean != OLD.ean) THEN
```
Comment (rschweizer) [inline CseEightselectBasic/CseEightselectBasic.php]: I might have missed it or am not well enough informed how Shopware and our plugin work in particular, but how do we know if any of the other values have changed, e.g. description, brand, color, ...?

File: CseEightselectBasic/Components/PropertyExport.php (hunk 2 above)
Comment (voodoods) [reply CseEightselectBasic/Components/PropertyExport.php]: `ES` stands for `EightSelect` and is merely a prefix for the environment variable `ES_DEBUG` so to not get confused with other debug outputs when you run i.e. the export in a terminal. `SQL_DEBUG` would not explicitly indicate that this debug output is coming from our plugin.

Comment (rschweizer) [reply CseEightselectBasic/Components/PropertyExport.php]: OK, in this case I would suggest being verbose and call it `EIGHTSELECT_DEBUG` or something really specific, because `ES` is often used as abbreviation for `elasticsearch`.

File: CseEightselectBasic/CseEightselectBasic.php (hunk 3 above)
Comment (voodoods) [reply CseEightselectBasic/CseEightselectBasic.php]: That's actually a good question 👍 There are some triggers missing here i.e. for brand, name, images, etc. - I will add them immediately. 

A color change (or for that matter any change in an option-article-relation) will be detected and logged by the `8s_s_article_configurator_option_relations_change_queue_writer` trigger on line 849.

**Comment by voodoods** [comment]
@rschweizer changes done, please re-check them 👍

File: CseEightselectBasic/CseEightselectBasic.php [hunk 4]
```diff
@@ -837,22 +836,90 @@ private function initChangesQueueTable()
...
+            'CREATE TRIGGER `8s_articles_change_queue_writer` AFTER UPDATE on `s_articles`
+                  FOR EACH ROW
+                  BEGIN
+                    IF (NEW.supplierID != OLD.supplierID OR 
+                    NEW.name != OLD.name OR 
+                    NEW.configurator_set_id != OLD.configurator_set_id OR 
+                    NEW.description != OLD.description OR 
+                    NEW.description_long != OLD.description_long) 
+                    THEN
+                      INSERT INTO 8s_articles_details_change_queue (s_articles_details_id, updated_at) VALUES (NEW.id, NOW());
```
Comment (Wytrykus) [inline CseEightselectBasic/CseEightselectBasic.php]: As discussed: This cannot work correctly because we're inserting an id from `s_articles` into `8s_articles_details_change_queue. s_articles_details_id `.
What we need instead is to insert a row into our changes queue for each product variant (i.e. for each row of `s_articles_details` that belongs to the current row of `s_articles`).

File: CseEightselectBasic/CseEightselectBasic.php [hunk 5]
```diff
@@ -866,9 +933,15 @@ private function initChangesQueueTable()
     private function deleteChangesQueueTable()
     {
         $triggerSqls = [
+            'DROP TABLE IF EXISTS `8s_articles_change_queue`;',
```
Comment (Wytrykus) [inline CseEightselectBasic/CseEightselectBasic.php]: This table is never created and never referenced --> obsolete

File: CseEightselectBasic/Components/PropertyExport.php [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+                s_articles_supplier.name AS marke,
+                s_core_tax.tax AS tax
+                FROM s_articles_details
+                    INNER JOIN s_articles ON s_articles.id = s_articles_details.articleID
+                    INNER JOIN s_articles_attributes ON s_articles_attributes.articledetailsID = s_articles_details.id
+                    INNER JOIN s_articles_prices ON s_articles_prices.articledetailsID = s_articles_details.id AND s_articles_prices.from = \'1\'
+                    INNER JOIN 8s_articles_details_change_queue ON 8s_articles_details_change_queue.s_articles_details_id = s_articles_details.id
+                    INNER JOIN s_articles_supplier ON s_articles_supplier.id = s_articles.supplierID
+                    INNER JOIN s_core_tax ON s_core_tax.id = s_articles.taxID
+                LIMIT ' . $number . ' OFFSET ' . $from;
```
Comment (Wytrykus) [inline CseEightselectBasic/Components/PropertyExport.php]: I don't understand how this can ever lead to a CSV file with all attributes, prices, suppliers etc. in one row - because the JOIN with `s_articles_attributes`, `s_articles_prices` etc. will create multiple rows for the same `s_articles_details.id` with all the different values found, and I don't see in the code where/how these are combined into one CSV line.

File: CseEightselectBasic/CseEightselectBasic.php [hunk 7]
```diff
@@ -837,22 +836,90 @@ private function initChangesQueueTable()
...
-                    IF (NEW.instock != OLD.instock OR NEW.active != OLD.active) THEN
-                      INSERT INTO 8s_articles_details_change_queue (s_articles_details_id, updated_at) VALUES (NEW.id, NOW());
+                    IF (NEW.articleID != OLD.articleID OR NEW.ordernumber != OLD.ordernumber OR NEW.instock != OLD.instock OR NEW.active != OLD.active OR NEW.ean != OLD.ean) THEN INSERT INTO 8s_articles_details_change_queue (s_articles_details_id, updated_at) VALUES (NEW.id, NOW());
                     END IF;
                   END',
+            'DROP TRIGGER IF EXISTS `8s_articles_img_change_queue_writer`',
+            'CREATE TRIGGER `8s_articles_img_change_queue_writer` AFTER UPDATE on `s_articles_img`
+            FOR EACH ROW
+            BEGIN
+            IF (NEW.img != OLD.img OR NEW.extension != OLD.extension) THEN INSERT INTO 8s_articles_details_change_queue (s_articles_details_id, updated_at) VALUES (NEW.articleID, NOW());
```
Comment (voodoods) [inline CseEightselectBasic/CseEightselectBasic.php]: In regards to CSE-1036 we have to listen for changes in the field `position` as well.

**Comment by voodoods** [comment]
@akleiber @rschweizer @Wytrykus This is ready to review now.

File: CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_manual_export/view/detail/export.js [hunk 8]
```diff
@@ -21,29 +21,26 @@ Ext.define("Shopware.apps.CseEightselectBasicManualExport.view.detail.Export", {
...
-        : "Letzter Voll-Export am: " + lastFullExportTimeStamp;
+        ? "Noch kein Voll-Export duchgeführt (alle Stammdaten)"
+        : "Letzter Voll-Export am: " + lastFullExportTimeStamp + " (alle Stammdaten)"
 
-    var lastQuickUpdateLabel =
-      lastQuickUpdateTimeStamp.length === 0
-        ? "Noch kein Schnell-Update durchgeführt."
-        : "Letztes Schnell-Update am: " + lastQuickUpdateTimeStamp;
+    var lastPropertyExportLabel =
+      lastPropertyExportTimeStamp.length === 0
```
Comment (rschweizer) [inline CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_manual_export/view/detail/export.js]: Why not simply have the ternary here?
```javascript
var lastPropertyExportLabel = lastPropertyExport
? "Noch keine Schnell-Update durchgeführt (nur Änderungen)"
: "Letztes Schnell-Update am: " + lastPropertyExportTimeStamp + " (nur Änderungen)";
```

https://github.com/8select/shopware-plugin-sob/pull/30/files#diff-1c524db20e9894c10717d01112c860e7R33 could then be removed. Same for the `lastFullExport`.

File: CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_manual_export/view/detail/export.js [hunk 9]
```diff
@@ -21,29 +21,26 @@ Ext.define("Shopware.apps.CseEightselectBasicManualExport.view.detail.Export", {
...
+        ? "Noch keine Schnell-Update durchgeführt (nur Änderungen)"
```
Comment (rschweizer) [inline CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_manual_export/view/detail/export.js]: Typo: 'keine' sollte 'kein' lauten.

File: CseEightselectBasic/Components/PropertyExport.php [hunk 10]
```diff
@@ -18,7 +18,7 @@ class PropertyExport extends Export
      */
     private $fieldMapping = [
         'prop_sku' => 'sku',
-        'prop_isInStock' => 'instock',
+        'prop_isInStock' => 'status',
```
Comment (rschweizer) [inline CseEightselectBasic/Components/PropertyExport.php]: https://github.com/8select/shopware-plugin-sob/pull/30/files#diff-ade3e3d0f029fa66db9b94eea6153b8dR900

File: CseEightselectBasic/Components/PropertyExport.php [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    const FEED_TYPE = 'property_feed';
+
+    const CRON_NAME = '8select_property_export';
+
+    /**
+     * @var array
+     */
+    private $fieldMapping = [
+        'prop_sku' => 'sku',
+        'prop_isInStock' => 'status',
```
Comment (rschweizer) [inline CseEightselectBasic/Components/PropertyExport.php]: https://github.com/8select/shopware-plugin-sob/pull/30/files#diff-ade3e3d0f029fa66db9b94eea6153b8dR900 sets a trigger on `instock`. Is that as intended? Should the trigger or the mapping be changed?

File: CseEightselectBasic/Components/Export.php [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+                $top = $numArticles;
+            }
+
+            if (getenv('ES_DEBUG')) {
+                echo sprintf('Processing articles %d to %d from %d%s', $i, $top, $numArticles, \PHP_EOL);
+            }
+
+            foreach ($articles as $article) {
+                $line = FieldHelper::getLine($article, $this->fields);
+                 $csvWriter->insertOne($line);
```
Comment (rschweizer) [inline CseEightselectBasic/Components/Export.php]: Indentation

**Comment by akleiber** [comment]
- trigger für s_articles_supplier hinzufügen
- roberts change requests
- installieren testen
- update von 1.5.2-RC1 auf 1.6.0-RC1 testen

File: CseEightselectBasic/Components/FeedLogger.php [hunk 13]
```diff
@@ -25,10 +25,20 @@ public static function logFeed($feedName) {
     Shopware()->Db()->query($sql);
   }
 
+  /**
+  * @param string $feedname 
+  * @throws \Zend_Db_Adapter_Exception
+  * @throws \Zend_Db_Statement_Exception
+  * @return string
```
[the remark this answers is not in the document]
Comment (voodoods) [reply CseEightselectBasic/Components/FeedLogger.php]: While I would argue that the within here used method `fetchOne()` does actually return the value of the record, see: https://framework.zend.com/manual/1.10/de/zend.db.adapter.html#zend.db.adapter.select.fetchone - which is a field of type `DateTime` which in PHP validates to `string`...

TL;DR
There's actually a type declaration for `DateTime` in Doctrine, which we already use within the bootstrap file: https://github.com/8select/shopware-plugin-sob/blob/master/CseEightselectBasic/CseEightselectBasic.php#L885

Changed it to that.

File: CseEightselectBasic/Components/PropertyExport.php (hunk 11 above)
Comment (voodoods) [reply CseEightselectBasic/Components/PropertyExport.php]: This is correct. Two things. Actually three things:

- `instock` and `active` are two properties that live independent of each other in Shopware; while an article may have no stock left (`instock === 0`) it can still be `active === true`, meaning it can be ordered regardless of stock; there is actually a sale-out flag ("Abverkauf") which declares an article to be set to `active === false` as soon as `instock` falls to `0`

- in the JIRA issue it is defined that want to have a boolean value here that tells is if product can be ordered or not; this happens to be the value of the field `status` in the full article export, which is mapped to the field `isInStock` here in the prop export - that may cause confusion

- as part of the trigger refactoring we ditched the trigger conditions altogether and just listen for all column changes in each of the affected tables

File: CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_manual_export/view/detail/export.js (hunk 8 above)
Comment (voodoods) [reply CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_manual_export/view/detail/export.js]: good point. done

File: CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_manual_export/view/detail/export.js (hunk 9 above)
Comment (voodoods) [reply CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_manual_export/view/detail/export.js]: done

File: CseEightselectBasic/Components/Export.php (hunk 12 above)
Comment (voodoods) [reply CseEightselectBasic/Components/Export.php]: done

**Comment by voodoods** [comment]
- [x] trigger für s_articles_supplier hinzugefügt
- [x] roberts change requests abgearbeitet
- [x] installation inkl. aller trigger und crons getestet
- [x] update von 1.5.2-RC1 auf 1.6.0-RC1 inkl. aller trigger und crons getestet
